src/scf/direct_minimization.jl

```diff
@@ -0,0 +1,132 @@
+# We do a bit of back and forth using custom `vec` (ours -> optim's) and `devec` (optim's -> ours) functions
+
+# Impose the orthogonality constraint on orbitals inside each k-block
+struct DMManifold <: Optim.Manifold
+    Nk::Int
+    devec_fun::Function
+end
+function Optim.project_tangent!(m::DMManifold, g, x)
+    g_devec = m.devec_fun(g)
+    x_devec = m.devec_fun(x)
+    for ik = 1:m.Nk
+        Optim.project_tangent!(Optim.Stiefel(),
+                               g_devec[ik],
+                               x_devec[ik])
+    end
+    g
+end
+function Optim.retract!(m::DMManifold, x)
+    x_devec = m.devec_fun(x)
+    for ik = 1:m.Nk
+        Optim.retract!(Optim.Stiefel(), x_devec[ik])
+    end
+    x
+end
+
+# Array of preconditioners
+struct DMPreconditioner
+    Nk::Int
+    Pks::Array # Pks[ik] is the preconditioner for kpoint ik
+    devec_fun::Function
+end
+function LinearAlgebra.ldiv!(p, P::DMPreconditioner, d)
+    p_devec = P.devec_fun(p)
+    d_devec = P.devec_fun(d)
+    for ik = 1:P.Nk
+        ldiv!(p_devec[ik], P.Pks[ik], d_devec[ik])
+    end
+    p
+end
+function LinearAlgebra.dot(x, P::DMPreconditioner, y)
+    x_devec = P.devec_fun(x)
+    y_devec = P.devec_fun(y)
+    sum(dot(x_devec[ik], P.Pks[ik], y_devec[ik])
+        for ik = 1:P.Nk)
+end
+function Optim.precondprep!(P::DMPreconditioner, x)
+    x_devec = P.devec
+    for ik = 1:P.Nk
+        precondprep!(Pks[ik], x_devec[ik])
+    end
+    P
+end
+
+
+"""
+Computes the ground state by direct minimization. `kwargs...` are passed to `Optim.Options()`.
+"""
+direct_minimization(basis::PlaneWaveBasis; kwargs...) = direct_minimization(basis, nothing; kwargs...)
+function direct_minimization(basis::PlaneWaveBasis{T}, Psi0;
+                             prec_type=PreconditionerTPA,
+                             optim_solver=Optim.LBFGS, kwargs...) where T
+    model = basis.model
+    @assert model.spin_polarisation in (:none, :spinless)
+    @assert model.assume_band_gap
+    filled_occ = filled_occupation(model)
+    n_bands = div(model.n_electrons, filled_occ)
+    ortho(Psik) = Matrix{Complex{T}}(Matrix(qr(Psik).Q))
+    Nk = length(basis.kpoints)
+
+    if Psi0 === nothing
+        Psi0 = [ortho(randn(Complex{T}, length(k.basis), n_bands)) for k in basis.kpoints]
+    end
+    occupation = [filled_occ*ones(T, n_bands) for ik = 1:Nk]
+
+    ham = Hamiltonian(basis)
+
+    ## vec and devec
+    # length of Psi[ik]
+    lengths = [length(Psi0[ik]) for ik = 1:Nk]
+    # psi[ik] is in psi_flat[starts[ik]:starts[ik]+lengths[ik]-1]
+    starts = copy(lengths)
+    starts[1] = 1
+    for ik = 1:Nk-1
+        starts[ik+1] = starts[ik] + lengths[ik]
+    end
+    vec(Psi) = vcat(Base.vec.(Psi)...) # TODO as an optimization, do that lazily? See LazyArrays
+    devec(Psi) = [@views reshape(Psi[starts[ik]:starts[ik]+lengths[ik]-1], size(Psi0[ik])) for ik in 1:Nk]
+
+    # this will get updated
+    energies = nothing
+
+    # computes energies and gradients
```

Comment:
Could be worth mentioning that `G` is to be updated in-place, but `E` not.

Reply:
That's standard optim interface

Reply:
Ok.